src/colmap/estimators/cost_functions.h

```diff
@@ -88,34 +145,42 @@ class ReprojErrorCostFunction {
                             &residuals[1]);
     residuals[0] -= T(observed_x_);
     residuals[1] -= T(observed_y_);
+    if constexpr (CTYPE != CovarianceType::IDENTITY) {
+      ApplySqrtInformation<T, 2, CTYPE>(residuals, sqrt_info_point2D_);
+    }
     return true;
   }
 
  private:
   const double observed_x_;
   const double observed_y_;
+  const Eigen::Matrix2d sqrt_info_point2D_;
```

Comment:
This could be conditionally defined based on whether we need the matrix or not:
```
struct IdentityCovariance {
};

struct DiagonalCovariance2d {
  Eigen::Vector2d cov;
};

struct DenseCovariance2d {
  Eigen::Matrix2d cover;
};

template<typename CTYPE>
struct CostFunction : std::conditional_t<std::is_same<CTYPE, CovarianceType::IDENTITY>::value, IdentityCovariance2d, std::conditional_t<std::is_same<CTYPE, CovarianceType::IDENTITY>::value, DiagonalCovariance2d, DenseCovariance2d>> {
  ...
};
```

Reply:
Probably doesn't compile right away, but you get the idea :-)

Reply:
Thanks. This is such a beautiful idea. I gave a try here: https://github.com/colmap/colmap/pull/2800/commits/6ec0074a168d67321fb06f6e2bb81a7b94cb1c3d. PTAL : )

Reply:
This will make pybind a little ugly though, as we dont have an interface with consistent type anymore. Also, those Covariance classes need to be binded, which feels a little complicated. Should we first revert to the stable one and try to support this feature in a separate PR?

Reply:
Reverted and put it here: https://github.com/B1ueber2y/colmap/tree/features/with_covariance_data_type for now. No clean way out for pybind with CovarianceDataType since to be able to support arbitrary CameraModel as template we need to do the argument forwarding. Will open a separate PR after this gets merged.

Reply:
I think the interface could still consistently take 2d matrix. But then we only store the relevant parts of it in the cost function members. However, maybe I don't understand some limitation of pybind that would still make this difficult.

Reply:
Ah OK if we make interfacing the same it would not be difficult, but I still suggest going for another PR for that change as I m not very sure about the complication : )

Reply:
Readded the support for Eigen Matrix. Squash and merge here: https://github.com/colmap/colmap/pull/2800/commits/13ce6a8300808cc256193b6383790fbcbb5cc78f. PTAL